src/txrequest.h

```diff
@@ -0,0 +1,332 @@
+#include <boost/multi_index/ordered_index.hpp>
+
+#include <chrono>
+#include <unordered_map>
+#include <vector>
+
+#include <stdint.h>
+
+
+/** A functor with embedded salt that computes priority of a txid/peer combination.
+ *
+ * The priority is used to order candidates for a given txid; lower priority values are selected first. It has the
+ * following properties:
+ * - Priorities for outbound peers are always lower than inbound peers, making them always preferred.
+ * - Within the set of outbound peers and within the set of inbound peers:
+ *   - The very first announcement for a given txid is always lowest.
+ *   - Announcements after the first one are ordered uniformly randomly (using a salted hash of txid/peer).
+ */
+class PriorityComputer {
+    const uint64_t m_k0, m_k1;
+public:
+    explicit PriorityComputer(bool deterministic);
+    uint64_t operator()(const uint256& txid, uint64_t peer, bool inbound, bool first) const
+    {
+        uint64_t low_bits = 0;
+        if (!first) {
+            low_bits = CSipHasher(m_k0, m_k1).Write(txid.begin(), txid.size()).Write(peer).Finalize() >> 1;
+        }
+        return low_bits | uint64_t(inbound) << 63;
+    }
+};
+
+/** Data structure to keep track of, and schedule, transaction downloads from peers.
+ *
+ * === High level behavior ===
+ *
+ * Transaction downloads are attempted in random order, first choosing among outbound peers, then inbound ones.
+ * Inbound peers are only considered after a (configurable) inbound delay, to give time to outbound peers. When a
+ * NOTFOUND is received, an invalid transaction is received, or a download times out (after a configurable delay),
+ * the next candidate is immediately scheduled according to the same rules above. The same transaction is never
+ * re-requested from the same peer, unless the transaction was forgotten about in the mean time. This happens
+ * whenever no candidates remain, or when a transaction is successfully received.
+ *
+ *
+ * === Specification ===
+ *
+ * The data structure maintains a collection of entries:
+ *
+ * - CANDIDATE (txid, peer, reqtime) entries representing txids that were announced by peer, and become available for
+ *   download after reqtime has passed (which is immediately for outbound peers, and inbound_delay after announcement
+ *   for inbound ones).
+ *
+ * - REQUESTED (txid, peer, exptime) entries representing txids that have been requested, and we're awaiting a
+ *   response for from that peer. The exptime value determines when the request times out.
+ *
+ * - COMPLETED (txid, peer) entries representing txids that have been requested from a peer, but they timed out, a
+ *   NOTFOUND message was received for them, or an invalid response was received. They're only kept around to prevent
+ *   downloading them again. If only COMPLETED entries exist for a given txid remain (so no CANDIDATE or REQUESTED
+ *   ones), all of them are deleted (this is an invariant, and maintained by all operations below).
```

Comment:
3d5b4f20015246b7a0dd74b2dc89f389204dd9b8

nanonit: the word "completed" associates with a "success" for me (but I'm not a native English speaker though). These entries are not processed successfully, i.e., a transaction has not been received. Maybe rename `COMPLETED` to more general `PROCESSED` or something similar?

Reply:
I've struggled with a good name for this, actually, and changed it a few times. The problem is that it corresponds to a number of scenarios:

* We've received the transaction (but we don't know (yet) if it's a valid one)
* We've received the transaction with an invalid witness
* We've received a NOTFOUND message for the transaction
* The request timed out

I don't think `PROCESSED` is an improvement, but I admit `COMPLETED` doesn't perfect either.